File: src/main/java/io/github/freshsupasulley/dbot/PoliticalParty.java
package io.github.freshsupasulley.dbot;

import java.util.HashSet;
import java.util.Set;

import net.dv8tion.jda.api.entities.Member;

// damn records aren't a thing in this java version :(
// getters and setters it is
public class PoliticalParty {
	
	private final long role, category;
	
	private Set<Long> blacklist = new HashSet<Long>();
	
	/** Leaders can change */
	private long leader;
	
	public PoliticalParty(long role, long category, long owner)
	{
		this.role = role;
		this.category = category;
		this.leader = owner;
	}
	
	/**
	 * Checks if the member is banned.
	 * 
	 * @return true if the member is banned, false otherwise
	 */
	public boolean isBanned(Member member)
	{
		return blacklist.contains(member.getIdLong());
	}
	
	/**
	 * Gets the ban list of the party.
	 * 
	 * @return list of all banned users
	 */
	public Set<Long> getBlacklist()
	{
		return blacklist;
	}
	
	/**
	 * Bans a member from this party.
	 * 
	 * @param member member
	 * @return true if the member was banned, false if already banned
	 */
	public boolean addBan(Member member)
	{
		return blacklist.add(member.getIdLong());
	}
	
	/**
	 * Unbans a member from this party.
	 * 
	 * @param member member
	 * @return true if the member was unbanned, false if was never banned
	 */
	public boolean removeBan(Member member)
	{
		return blacklist.remove(member.getIdLong());
	}
	
	public long getRole()
	{
		return role;
	}
	
	public long getCategory()
	{
		return category;
	}
	
	public long getLeaderID()
	{
		return leader;
	}
	
	public void setLeader(long leader)
	{
		this.leader = leader;
	}
	
	public boolean equals(PoliticalParty party)
	{
		return party != null && party.role == this.role;
	}
}

File: src/political-party.ts
export class PoliticalParty {
	// Also serves as the party ID
	private role: string;
	private leader: string;

	private blacklist: string[] = [];

	constructor(role: string, leader: string) {
		this.role = role;
		this.leader = leader;
	}

	public isBanned(memberID: string): boolean {
		return this.blacklist.includes(memberID);
	}

	public getBlacklist(): string[] {
		return this.blacklist;
	}

	/**
	 * Adds this user to the party blacklist.
	 *
	 * Does **NOT** perform any API calls.
	 *
	 * @param member Discord user ID
	 * @returns true if user was blacklisted, false if already banned
	 */
	public addBan(member: string): boolean {
		if (this.isBanned(member)) return false;
		this.blacklist.push(member);
		return true;
	}

	public removeBan(member: string): boolean {
		const index = this.blacklist.indexOf(member);
		if (index !== -1) {
			this.blacklist.splice(index, 1);
			return true;
		}
		return false;
	}

	/**
	 * Gets the role ID of this party.
	 * @returns the Discord role ID of this party
	 */
	public getID(): string {
		return this.role;
	}

	public getLeaderID(): string {
		return this.leader;
	}

	public setLeaderID(memberID: string) {
		this.leader = memberID;
	}

	public equals(party: PoliticalParty | null): boolean {
		return party !== null && party.getID() === this.role;
	}
}
